File: python/stage-4/level-1/main.py
#!/usr/bin/env python3
"""
Temperature Converter
A simple Python program for converting between Celsius and Fahrenheit.
"""

def celsius_to_fahrenheit(celsius):
    """Convert Celsius to Fahrenheit."""
    return (celsius * 9/5) + 32

def fahrenheit_to_celsius(fahrenheit):
    """Convert Fahrenheit to Celsius."""
    return (fahrenheit - 32) * 5/9

def get_temperature_input(prompt):
    """Get temperature input from user with validation."""
    while True:
        try:
            temp = float(input(prompt))
            return temp
        except ValueError:
            print("❌ Invalid input. Please enter a valid number.")

def get_menu_choice():
    """Get menu choice from user with validation."""
    while True:
        try:
            choice = int(input("Enter choice (1 or 2): "))
            if choice in [1, 2]:
                return choice
            else:
                print("❌ Please enter 1 or 2.")
        except ValueError:
            print("❌ Invalid input. Please enter a number.")

def main():
    """Main program function."""
    print("🌡️  Temperature Converter 🌡️")
    print("=" * 30)
    print("What conversion do you want?")
    print("1. Celsius to Fahrenheit")
    print("2. Fahrenheit to Celsius")

    choice = get_menu_choice()

    if choice == 1:
        # Celsius to Fahrenheit
        celsius = get_temperature_input("Enter temperature in Celsius: ")
        fahrenheit = celsius_to_fahrenheit(celsius)
        print(".1f")
    elif choice == 2:
        # Fahrenheit to Celsius
        fahrenheit = get_temperature_input("Enter temperature in Fahrenheit: ")
        celsius = fahrenheit_to_celsius(fahrenheit)
        print(".1f")
if __name__ == "__main__":
    main()

File: rust/stage-4/level-1/main.rs
use std::io::{self, Write};

fn main() {
    println!("Temperature Converter");
    println!("=====================");
    println!("What conversion do you want?");
    println!("1. Celsius to Fahrenheit");
    println!("2. Fahrenheit to Celsius");
    print!("Enter choice (1 or 2): ");

    // Flush stdout to ensure prompt appears
    io::stdout().flush().unwrap();

    let mut choice = String::new();
    io::stdin().read_line(&mut choice).unwrap();
    let choice: u32 = match choice.trim().parse() {
        Ok(num) => num,
        Err(_) => {
            println!("Invalid choice. Please enter 1 or 2.");
            return;
        }
    };

    if choice == 1 {
        // Celsius to Fahrenheit
        print!("Enter temperature in Celsius: ");
        io::stdout().flush().unwrap();

        let mut celsius = String::new();
        io::stdin().read_line(&mut celsius).unwrap();
        let celsius: f64 = match celsius.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("Invalid temperature. Please enter a number.");
                return;
            }
        };

        let fahrenheit = (celsius * 9.0 / 5.0) + 32.0;
        println!("{:.1}°C is equal to {:.1}°F", celsius, fahrenheit);

    } else if choice == 2 {
        // Fahrenheit to Celsius
        print!("Enter temperature in Fahrenheit: ");
        io::stdout().flush().unwrap();

        let mut fahrenheit = String::new();
        io::stdin().read_line(&mut fahrenheit).unwrap();
        let fahrenheit: f64 = match fahrenheit.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("Invalid temperature. Please enter a number.");
                return;
            }
        };

        let celsius = (fahrenheit - 32.0) * 5.0 / 9.0;
        println!("{:.1}°F is equal to {:.1}°C", fahrenheit, celsius);

    } else {
        println!("Invalid choice. Please enter 1 or 2.");
    }
}